rotkehlchen/exchanges/okx.py

```diff
@@ -0,0 +1,498 @@
+        url = urljoin(self.base_uri, path)
+
+        try:
+            response = self.session.request(method=method, url=url)
+        except requests.exceptions.RequestException as e:
+            raise RemoteError(f'{self.name} API request failed due to {str(e)}') from e
+        try:
+            json_response = response.json()
+        except requests.exceptions.JSONDecodeError as e:
+            raise RemoteError(
+                f'{self.name} returned invalid JSON response: {response.text}',
+            ) from e
+
+        return json_response
+
+    def _api_query_list(
+            self,
+            endpoint: Literal['currencies', 'balance', 'trades', 'deposits', 'withdrawals'],
+            options: Optional[dict] = None,
+    ) -> list:
+        """
+        Makes an API query and parses the response into a list
+
+        May raise
+        - RemoteError if the API returns an unexpected response
+        """
+        response = self._api_query(endpoint=endpoint, options=options)
+        data = response.get('data')
+        if data is None or not isinstance(data, list):
+            raise RemoteError(
+                f'{self.name} json response does not contain list `data`: {response}',
+            )
+        return data
+
+    def _api_query_list_paginated(
+            self,
+            endpoint: Literal['currencies', 'balance', 'trades', 'deposits', 'withdrawals'],
+            pagination_key: str,
+            options: Optional[dict] = None,
+    ) -> list:
+        """
+        Makes subsequent API queries until response list length is less than MAX_RESULTS.
+        Pagination is handled by setting `pagination_key` value in the `after` query parameter.
+        This returns items with `pagination_key` value less than the passed value
+
+        May raise
+        - RemoteError if the API returns an unexpected response
+        """
+        options = options.copy() if options else {}
+        all_items = []
+
+        while True:
+            data = self._api_query_list(endpoint=endpoint, options=options)
+            all_items.extend(data)
+            if len(data) < self.MAX_RESULTS:
+                break
+
+            earliest_item = data[-1]
+            options.update({'after': earliest_item[pagination_key]})
+
+        return all_items
+
+    def validate_api_key(self) -> tuple[bool, str]:
+        response = self._api_query(endpoint='balance')
+        if response.get('code') != '0':
+            return False, 'Error validating API credentials'
+        return True, ''
+
+    def query_balances(self, **kwargs: Any) -> ExchangeQueryBalances:
+        """
+        https://www.okx.com/docs-v5/en/#rest-api-account-get-balance
+
+        May raise
+        - RemoteError if the OKX API or price oracle returns an unexpected response
+        """
+        data = self._api_query_list(endpoint='balance')
+        try:
+            currencies_data = data[0]['details']
+        except (IndexError, KeyError, TypeError) as e:
```

Comment:
continuing the `TypeError` discussion do you see the problem here?

Such a `TypeError` would read: `"TypeError: string indices must be integers"`

Which would translate to: "OKX balance API request failed due to unexpected response TypeError: string indices must be integers".

We won't have any information as to what went wrong or which attribute it is that failed for us.

I think it's much better if you check that `data[0]` is a `dict` here and if not raise a `RemoteError` saying that the `"details"` key of the OKX balances response was not a dict as expected, and even include the raw response so it can be studied in the logs.

What we usually do with this is:
1. Put as few info as needed in the `RemoteError` as this is usually propagated high up, caught by the API and shown to the user.
2. At the same time log an error, with the raw response `data` so that when we study the logs of the user we can see both the error and the raw response and figure out what happened.

Reply:
You missed to do (2).

Logging the error containing the raw response so we, the developers, can debug it.